### PR TITLE
[4.0] [MVC] Always set model/view prefix, correct base path detection

### PR DESCRIPTION
**Pull request by SharkyKZ**
Pull Request for Issue #23510.

### Summary of Changes

This should fix using admin base path in frontend, e.g. when using Editors-XTD plugins.

### Testing Instructions

Edit an article in frontend.
In editor, click CMS Content and any button (except Page Break or Read More)

### Expected result

Modal with a list of items opens up.

### Actual result

Modal with error appears, e.g.:

     View not found [name, type, prefix]: contacts, html, 

### Documentation Changes Required

Maybe.

@mbabker @laoneo @wilsonge please review.

### CODE REVIEW

**Comment by mbabker** [comment]
All this pull request is doing is re-affirming my feeling that it is time to deprecate having separate extensions per application (minus templates).  Because there are more hacks to "proxy" things than benefits this separation really brings.

**Comment by SharkyKZ** [comment]
If this is about the messy code, it seems to be here for B/C-ish reasons. We could stop using `base_path` configuration option in J4 components and just pass the appropriate prefix everywhere. Similarly like suggested here https://github.com/joomla/joomla-cms/pull/23539#issuecomment-454037742.

File: libraries/src/MVC/Controller/BaseController.php [hunk 1]
```diff
@@ -755,9 +745,21 @@ public function getModel($name = '', $prefix = '', $config = array())
...
 
-		if (empty($prefix) && $this->factory instanceof LegacyFactory)
+		if (!$prefix)
 		{
-			$prefix = $this->model_prefix;
+			if ($this->factory instanceof LegacyFactory)
+			{
+				$prefix = $this->model_prefix;
+			}
+			// When the front uses a back end model
```
Comment (brianteeman) [inline libraries/src/MVC/Controller/BaseController.php]:
```suggestion
			// When the frontend uses an administrator model
```

File: libraries/src/MVC/Controller/BaseController.php [hunk 2]
```diff
@@ -865,9 +867,21 @@ public function getView($name = '', $type = '', $prefix = '', $config = array())
...
 
-		if (empty($prefix) && $this->factory instanceof LegacyFactory)
+		if (!$prefix)
 		{
-			$prefix = $this->getName() . 'View';
+			if ($this->factory instanceof LegacyFactory)
+			{
+				$prefix = $this->getName() . 'View';
+			}
+			// When the front uses a back end view
```
Comment (brianteeman) [inline libraries/src/MVC/Controller/BaseController.php]:
```suggestion
			// When the front uses an administrator view
```

**Comment by twister65** [comment]
I have tested this item :white_check_mark: successfully on cb8d6cc1d279725d35d9157fa17bf0a06173a584<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25367">issues.joomla.org/tracker/joomla-cms/25367</a>.</sub>

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on cb8d6cc1d279725d35d9157fa17bf0a06173a584<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25367">issues.joomla.org/tracker/joomla-cms/25367</a>.</sub>

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on cb8d6cc1d279725d35d9157fa17bf0a06173a584<br /><br />It seems that some backend language filesa are not loaded in the modals, but I think that's another issue and not subject of this PR, right?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25367">issues.joomla.org/tracker/joomla-cms/25367</a>.</sub>

**Comment by alikon** [comment]
i suppose so

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25367">issues.joomla.org/tracker/joomla-cms/25367</a>.</sub>

**Comment by SharkyKZ** [comment]
@richard67, yes. Related issue #25344.

**Comment by richard67** [comment]
@SharkyKZ Ha, I knew there was something like that. Other question. Are you still waiting for requested reviews, or are the good tests sufficient for you?

**Comment by SharkyKZ** [comment]
It's up to maintainers now. Though I agree with @mbabker this is a rather ugly hack. We can probably do better here.

**Comment by richard67** [comment]
Sure, but as far as I understand him, the right way would be far beyond the scope of this PR, and so I think this "ugly hack" is ok as a bug fix right now. Medium or long term we should really think over if this admin site thing still makes sense for plugins, modules, packages, and components, if I understand him right.

**Comment by richard67** [comment]
We already have other ugly hacks for the same reason, see e.g. my PR #25404 .

**Comment by SharkyKZ** [comment]
I don't think that PR was a hack. But even if it was, it's not an excuse to add more hacks.

**Comment by richard67** [comment]
Depends ... if adding hacks means more consistency because existing code is alrerady full of hacks it could make sense 😄 (attention, sarcastic joke)

**Comment by SharkyKZ** [comment]
>Sure, but as far as I understand him, the right way would be far beyond the scope of this PR

We could do it here. This PR really just moves some existing code around. The code that was initially added for B/C reasons, so we wouldn't have to set prefixes per component. We could remove this code from MVC library and set correct prefixes in components. But this would affect 3rd party components as they would also have to do the same.

**Comment by richard67** [comment]
I see.

**Comment by wilsonge** [comment]
Thanks!

Actually spent some time on this and the fix looks cleaner to me than what we had before too

**Pull request merged**